[zzawook] iP

zzawook opened the pull request:
* Dongji
> Dongji is a Korean word 'friend' that knows you well, and have known you for years. This application will be on your side, as you live your busy, productive life.

Dongji is an todo list application, that helps you remember things you need to do. It is:

- command-line based
- easy to learn
- SUPER FAST to use

To try it for yourself,

1. download the jar file from [here](https://github.com/zzawook/ip/releases/tag/v0.2.0)
2. Open it with the command
```
java -jar dongji.jar
```
3. add your tasks
4. let it manage your tasks for you

And it's FREE!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

zzawook committed:
Renamed the project to "Dongji", which means a friend in Korean, and changed all reference to Duke correspondingly

Added exit message, and separated greeting and exit message to separate function

zzawook committed:
Added echo feature to echo the user input

Created class declaration Dongji and moved all chatbot functionalities to object level declaration

zzawook committed:
Added 'add task' feature and 'list task' feature as new functions

Commented out unused 'echo' function

zzawook committed:
Encapsulated task as new Task class, and defined its behaviors

Organized the printing messages

zzawook committed:
Added 3 types of new Tasks (Event, Todo, Deadline) as subclass of Task, and muted Tasks as abstract.

Implemented corresponding toString behaviors according to property of each type.

zzawook committed:
Implemented error-handling using Exceptions.

Created custom exceptions to handle exceptions encountered in Dongji project

zzawook committed:
Added 'delete' feature

Added new exception that handles deleting, marking, and unmarking invalid task index.

zzawook committed:
Resolved compilation error in previous version

zzawook committed:
Created static factory of each task type and moved all object generation logic to each respective task type's class.

zzawook committed:
Added semi-automated testing using shell script

Improved Deadline's exception message

zzawook committed:
Refactored Dongji class by creating TaskList and encapsulating list related operations into TaskList class

Changed to use default runtest.sh provided from scaffold code given.

zzawook committed:
Refactor the code to use higher level of OOP

Classes in the project, especially the Dongji class, are undertaking several roles and "reasons to change" thus going against Single Responsibility Principle.

The design has low scalability and will make feature addition and future changes difficult.

Introduce new component 'Parser' to undertake parsing of the string command and generate corresponding Command instance. Introduce the Command interface and its implementations to represent Command internally, thus take advantage of polymorphism. Introduce Ui component which undertakes displaying of informations.

Dividing the role of Dongji to only undertake the entrance point role will improve scalabilitty and maintainability by dividing the high level logic (Command, Task) and low level logic (Ui) in separate component. This will also make the code in Dongji class more simple and legible as well.

Let's,
* Separate display logic into Ui component
* Separate parsing logic into Parser component
* Separate command specific execution details into Command interface and its implementations

zzawook committed:
Add new feature: Saving of working tasks

When the application terminates, all the task data user is working on gets deleted because the application didn't have persistence layer.

For flawless user experience, the task should be able to persist even after the application terminates, so that when user comes back they can continue working on their tasks.

Add interface Persistence that defines how persistence layer should behave, and an implentation Txt that defines logics to save onto .txt file format.

Programming on interface Persistence will enable future addition of saving on other file formats, which grant additional scalability and flexibility. Adding Txt implementation will grant basic savings functionality based on Persistence interface without losing significant feature at the moment.

Let's add Saving feature by adding Persistence interface and its implementation Txt.

zzawook committed:
Use LocalDateTime instead of String for deadlines and events

The deadline and event classes are using Strings instead of formal date/time types.

This could be misleading because anything can go into date/time field, which does not make sense in common use cases.

Replace all occurrences of String dates into DateTimeData, which is a wrapper around LocalDateTime with options to whether it would contain time as well.

Using formal DateTime instead of String for dates would prevent users inputting invalid values in date field. Additionally, using a wrapper DateTimeData would allow users flexibility to choose whether to use time together with date or not.

Let's replace String date formats to DateTimeData to grant date format safety and users' flexibility.

zzawook committed:
EventCommand: prevent event's end date being earlier than start date

The EventCommand allows the user to input end date earlier than start date, which does not make sense and goes against normal use case.

Add check to prevent event's end date being earlier than start date

By adding a check, we can prevent user adding an event with wrong timeframe by mistake (e.g. typo).

Let's check the validity of start and end date of event by adding a check.

zzawook committed:
Package the application into Dongji package

The application was in default directory, and not packaged.

Create new package Dongji and move the entire project into it.

Let's move the application into valid package.

zzawook committed:
Merged gradle support branch

zzawook committed:
Add test for ListCommand class and Event class

zzawook committed:
Modify build.gradle vscode's settings.json to build the JAR using gradle

zzawook committed:
Add Javadoc comments to public methods

zzawook committed:
Add Feature: Find task by keyword

Dongji does not support find task by keyword.

When task list gets exhaustive and the user cannot see all tasks at once, find by keyword is necessary, but Dongji doesn't know how to do it.

Add Find task by keyword, by creating FindCommand and implementing finding logics within. Add 'find' command parsing to CommandParser so that it can understand find command.

A new component FindCommand is necessary to represent new type of Command. Then, update Parser so that it can  identify "find" from the user input.

Let's add FindCommand and modify CommandParser to enable 'find by keyword' feature

kikokidayo commented on src/main/java/Dongji/components/DateTimeData.java:
Perhaps can consider renaming the boolean method `getHasTime` to be `hasTime` instead to make it sound more like a boolean method?

kikokidayo commented on src/main/java/Dongji/components/tasks/Task.java:
Could consider writing Javadoc comments for some public classes

kikokidayo commented on src/main/java/Dongji/components/parsers/CommandParser.java:
Nice job on formatting the switch statement!

kikokidayo commented on src/main/java/Dongji/components/commands/TodoCommand.java:
Could consider adding a short description/summary of the method `execute`

kikokidayo commented on src/main/java/Dongji/components/persistences/Txt.java:
Just a small nitpick but there should be a spacing between the `catch` command and the opening parenthesis on line 48 and 67

kikokidayo reviewed (commented):
Overall the code standard looks good! Just consider making some small minor syntax changes and adding some more Javadoc comments for not only methods but class names as well.

zzawook committed:
Add checkstyle to check Java code against 2103T coding standards

There is no explicit tool to check coding style and keep it under control.

This could lead to mistaken coding style that does not obey the coding standards set by the teaching team.

Adding checkstyle plugin to the project to take advantage of static analysis.

Let's add checkstyle to the project, to prevent the code becoming a spaghetti, with the help of static analysis.

zzawook committed:
Add GUI on the application using JavaFX

The application support text-based interaction with the user.

This could be difficult for normal users without much affinity to keyboard to use the application, thus decreasing generality of application.

Adding GUI to the application using JavaFX.

The implementation of GUI is not perfect yet, as the implementation only echoes the user input.

Let's add basic GUI setup to the project to lay foundation for further development of GUI on the project.

zzawook committed:
Connect Dongji backend with JavaFx frontend

The application only can echo the command entered by the user.

This does not match the specification for application, which is an interactive todo list.

Connecting Dongji backend with JavaFx Frontend by extracting previously frontend specific behaviors from Dongji class into CommandLine class, and isolating it as fully backend component, not having to know about frontend logics. Creating new Main class to act as launcher to launch either CommandLine application, or GUI application.

The GUI is now responding with same text as CommandLine interface. This lowers the barrier of using the application.

The modified design improves design of the application because the high-level component (Dongji) doesn't know anything about low-level component, thus cleaning up the wrong dependency directions. This will allow greater flexibility in the application.

Let's complete GUI implementation by connecting Dongji backend and JavaFx frontend, and cleaning up dependency directions.

mrnobody0505 commented on src/main/java/Dongji/components/tasks/Task.java:
```diff
@@ -0,0 +1,52 @@
+    /** 
+     * @return boolean
+     */
+    public boolean getIsMarked() {
```
Consider rename the function to isMarked() so that is has a clearer boolean meaning

mrnobody0505 commented on src/main/java/Dongji/components/ui/CommandLine.java:
```diff
@@ -0,0 +1,62 @@
+import Dongji.Dongji;
+
+public class CommandLine {
+    private final String SEPARATOR = "-------------------------------";
```
just a minor stuff but maybe this field should be private static final ?

mrnobody0505 commented on src/main/resources/view/MainWindow.fxml:
```diff
@@ -0,0 +1,43 @@
+            </content>
+        </ScrollPane>
+    </children>
+</AnchorPane>
```
Good works on fxml tho

zzawook committed:
Revise the Code to obey Coding Standard as suggested by peer reviews

zzawook committed:
Revise the code to obey Coding Standard as suggested by peer review

zzawook committed:
Use assert feature to document assumptions in the code

The application uses Exceptions to check on expected errors by the user or developer.

However, there may be unexpected errors, especially due to misled or  missed out assumptions in the code, which might introduce unexpected bugs to the program, thus leading the program to collapse.

Adding assertions to document important assumptions that should hold at various points in the code.

The assertion was used to
1. Valide input parameters for private methods
2. Points in the program to ensure validity of a fact which is almost certainly true
3. Validating post condition of the method, before returning the result

Assertions will help the developer detect the bug in development phase, thus preventing the bug to slip-through and cause issue later on in the dev process. It helps finding out problems at earlier stage of development as possible, thus decreasing cost.

Let's add assertions to document assumptions that must hold true within the program.

zzawook committed:
Refactor code to comply with coding standard using CheckStyle

zzawook committed:
Improve code quality abiding to coding standard proposed by 2103 team

The code have some minor coding style nitpicks that weren't resolved, such as opening brackets not following a white space, catch not followed by white space, etc.

Such deviations from coding standard can lead to challenges in collaboration on projects in any given time in the future.

Improving codes that deviate from coding standards to follow coding standard proposed by the 2103 team.

Improved code quality will decrease technical debt and enable greater efficiency in case of collaborative development in the future.

Let's improve code quality by changing deviations from code standard.

zzawook committed:
Merge pull request #1 from zzawook/branch-A-CodeQuality

Branch a code quality

zzawook committed:
Use streams instread of simple loops

While loops are effective way to iterate over a collection, streams are also another viable option to iterate over a collection.

Using streams instead of simple loops in find command and list command.

Though the changes does not bring performance optimization, practicing using Streams in this project can help me familiarize with it, in the case I might need them in the future.

Let's use streams instead of loops to iterate over collection.

zzawook committed:
Remove unnecessary dependency in FindCommand

zzawook committed:
Merge branch 'master' into branch-A-Streams

zzawook committed:
Merge pull request #2 from zzawook/branch-A-Streams

Branch a streams

zzawook committed:
Configure GitHub Action to build the jar file and automatically generate release

Donji, when a new version is released, or a new feature is added, the changes are not immediately reflected to the latest release. Rather, the developer needs to manually create a new jar file, and attach to a new release.

This can be cumbersome, and does not align with fast delivery to reflect user feedback swiftly, which is crucial to survive in software industry.

Adding Github Action action to automate build and release.

Addition of CI into the project will enable the user of this product to enjoy latest product at anytime, and delivery newest features as soon as it is implemented on the code.

Let's use Github Action to automate build and release.

zzawook committed:
Mute style and testing check for now to validate github action yml file

zzawook committed:
Debug use of github.ref

zzawook committed:
Merge pull request #3 from zzawook/branch-A-CI

Add GitHub Action to automate build and deployment

zzawook committed:
debug class not found error due to changed package name

zzawook committed:
Debug build.gradle file to reflect main class changed from GUI implementation

zzawook committed:
Prevent new release create a tag ending with
master, thus preventing push to origin/master

zzawook committed:
Add feature: Recurring tasks

The application can handle Todo, Event, Deadline.

However to represent repeating event, the user have to add new task every repeating period, which is cumbersome.

Adding new feature of recurring tasks, that uses crontab expression to express recurrence.

Cron is a universal expression that shows repetition. Thanks to its flexibility, we can express virtually every type of repetition we can imagine.

By adding recurring task, the user will be able to reduce the cumbersomeness of creating recurring event every time.

Let's add recurring task to represent repeating events.

zzawook committed:
Configure github action to be triggered only on master branch change

zzawook committed:
Merge tag 'master' of https://github.com/zzawook/ip

zzawook committed:
Merge branch 'master' of https://github.com/zzawook/ip

zzawook committed:
Improve GUI by adding a launcher class and a polishing up the UI

The application has an awkward class structure, which the role of Main
class is ambiguous. Also, the user interface of current GUI is not
intuitive for user, and lacks clarity in communicating the intention of
the application.

Add a launcher class to improve the structure of the application, and
polish up the UI to make it more intuitive and user-friendly.

The launcher class is a new class that is responsible for launching the
application. It is the entry point of the application and is responsible
for setting up the main window and initializing the UI components.

The UI is polished up by making the following changes:
- Different color for user and Dongji messages
- Improved layout for better readability
- Red message that alerts errors and warnings

Let's add a launcher class to improve the structure of the application,
and polish up the UI to make it more intuitive and user-friendly.

zzawook committed:
Apply minor changes to DialogBox.java, DialogBox.fxml

Give more padding to the icons

zzawook committed:
Change the images for the user and the duke

The image of Dongji and the user was a default image that had less
correlation with what the actual user case context was.

Change the images to be more representative of the user and the Dongji's
nature of the application.

The changed images are a better representation of user and Dongji.

zzawook committed:
Show welcome and goodbye messages

Dongji remains silent on launch, and cannot exit the application on
command. This is not user-friendly.

Add welcome and goodbye messages

Dongji now greets the user with a welcome message upon launch, and
can exit the application with the 'bye' command.

This is more user-friendly and intuitive.

zzawook committed:
Add more Javadoc comments to the codebase

zzawook committed:
Fixe bug in TaskList that listTasks fail when list is empty

zzawook committed:
Add screenshot of the UI

zzawook committed:
Add help command

Dongji cannot be used without knowing the commands.
However, the commands are not listed anywhere, so
users have to guess the commands or read from the
user guide.

Add a help command that lists all the commands available in Dongji.

Adding help command that lists all the commands will
help users find out how to use the application within
the application itself.

Let's add help command to the application for users
convenience.

zzawook committed:
Revise README to contain valid contents

The application does not have any documentation for users to find out
how the application is working, and user doesn't know how to interact
with Dongji.

Revise the current README.md file to explain how to use the application.

The README.md will help users to understand how to use the application
more intuitively using human-friendly languages and examples.

zzawook committed:
Revise README.md to improve readability

zzawook committed:
Add Table of Content to README.md

zzawook committed:
Add divider after each feature introduction

zzawook committed:
Add Testings for various classes in the project

The project does not have extensive testings for the entirity classes,
and does not ensure high coverage for the entire code base.

Add testings for the following classes:
- DeadlineCommand
- EventCommand
- ListCommand
- TodoCommand
- CommandParser

More rigorous testing will help to ensure the correctness of the
project, but requires more testings to be added in the future.

zzawook committed:
Modify the files to fix the checkstyle errors

Applying the checkstyle rules to the files

zzawook committed:
Fix minor bugs due to Java version differences

zzawook committed:
Attempt to delete release with same name

zzawook committed:
Update gradle.yml

zzawook committed:
Change checkstyle.xml to fit course's coding style standard

zzawook committed:
Apply more revisions to code to comply checkstyle

Apply revisions especially on test codes.
Also, roll back the GitHub Action workflow to
test the application before release.

zzawook committed:
Add more tests to cover most of the committed

Add tests to most non-trivial component-level
classes to ensure that they are working as expected.

zzawook committed:
Apply CheckStyle fixes for changes due to new test

zzawook committed:
Minor coding style issue fix

zzawook committed:
Fix integrity of the tests

zzawook committed:
Update build.gradle to point to  correct Launcher class

zzawook committed:
Make @AfterAll method in test class static
